File: backend/app/repositories/properties.py
from typing import Any

from motor.motor_asyncio import AsyncIOMotorDatabase

from app.models.property import PropertySearchFilters
from app.repositories.base import MongoRepository


class PropertyRepository(MongoRepository):
    collection_name = "properties"

    def __init__(self, db: AsyncIOMotorDatabase):
        self.db = db
        super().__init__(db[self.collection_name])

    async def search(self, filters: PropertySearchFilters, limit: int = 40) -> list[dict[str, Any]]:
        query: dict[str, Any] = {}
        if filters.budget_max:
            query["rent"] = {"$lte": filters.budget_max}
        if filters.property_types:
            query["property_type"] = {"$in": filters.property_types}
        if filters.locality_ids:
            query["locality_id"] = {"$in": filters.locality_ids}
        if filters.amenities:
            query["amenities"] = {"$all": filters.amenities}

        collections = await self._property_collections()
        per_collection_limit = max(limit, 100)
        docs: list[dict[str, Any]] = []

        for coll in collections:
            cursor = coll.find(query).sort("rent", 1).limit(per_collection_limit)
            docs.extend(await cursor.to_list(length=per_collection_limit))

        docs.sort(key=lambda item: item.get("rent", 0))
        return docs[:limit]

    async def _property_collections(self) -> list[Any]:
        names = await self.db.list_collection_names()
        property_names = [name for name in names if name == self.collection_name or name.startswith("properties_")]
        if not property_names:
            return [self.collection]
        return [self.db[name] for name in sorted(property_names)]

    async def upsert_by_dedupe_key(self, payload: dict[str, Any]) -> dict[str, Any]:
        await self.collection.update_one(
            {"dedupe_key": payload["dedupe_key"]},
            {"$set": payload, "$setOnInsert": {"created_at": payload.get("created_at")}},
            upsert=True,
        )
        return await self.collection.find_one({"dedupe_key": payload["dedupe_key"]}) or payload

File: backend/src/repositories/properties.ts
import { Db, Filter } from "mongodb";
import { PropertySearchFilters } from "../models/property.js";
import { MongoRepository, serializeDoc } from "./base.js";
import { cityPartition } from "../data/multiCity.js";

export class PropertyRepository extends MongoRepository {
  protected db:Db; public static readonly COLLECTION_NAME="properties";
  constructor(db:Db){super(db.collection(PropertyRepository.COLLECTION_NAME));this.db=db;}

  async search(filters:PropertySearchFilters,limit:number=40):Promise<any[]>{
    const query:Filter<any>={};if(filters.budget_max)query.rent={$lte:filters.budget_max};if(filters.property_types?.length)query.property_type={$in:filters.property_types};if(filters.locality_ids?.length)query.locality_id={$in:filters.locality_ids};if(filters.amenities?.length)query.amenities={$all:filters.amenities};if(filters.city)query.city={$regex:new RegExp(`^${filters.city.replace(/[.*+?^${}()|[\]\\]/g,"\\$&")}$`,`i`)};
    const collections=await this.getPropertyCollections(filters.city||undefined);const perCollectionLimit=Math.max(limit,100);const docs:any[]=[];
    for(const coll of collections){const results=await coll.find(query).sort({rent:1}).limit(perCollectionLimit).toArray();docs.push(...results);}
    docs.sort((a,b)=>(a.rent||0)-(b.rent||0));return docs.slice(0,limit).map(serializeDoc);
  }

  async getPropertyCollections(city?:string):Promise<any[]>{
    const collections=await this.db.listCollections().toArray();const names=collections.map((c)=>c.name);
    if(city){const partition=cityPartition(city);const cityCollections=names.filter((name)=>name===partition);if(cityCollections.length)return cityCollections.map((name)=>this.db.collection(name));}
    const propertyCollectionNames=names.filter((name)=>name===PropertyRepository.COLLECTION_NAME||name.startsWith("properties_")).sort();
    if(propertyCollectionNames.length===0)return[this.collection];return propertyCollectionNames.map((name)=>this.db.collection(name));
  }

  async upsertByDedupeKey(payload:Record<string,any>):Promise<any>{const now=new Date();const{created_at,...updateFields}=payload;await this.collection.updateOne({dedupe_key:payload.dedupe_key},{$set:updateFields,$setOnInsert:{created_at:created_at||now}},{upsert:true});const doc=await this.collection.findOne({dedupe_key:payload.dedupe_key});return serializeDoc(doc)||payload;}
}
